packages/core/src/external/carequality/pd/process-incoming-pd.ts

```diff
@@ -0,0 +1,135 @@
+import { MPI } from "../../../mpi/mpi";
+import { patientMPIToPartialPatient } from "../../../mpi/shared";
+import { toFHIR as convertPatientToFHIR } from "../../fhir/patient";
+import {
+  InternalError,
+  LivingSubjectAdministrativeGenderRequestedError,
+  PatientAddressRequestedError,
+  validateFHIRAndExtractPatient,
+} from "./validating-pd";
+
+const METRIPORT_HOME_COMMUNITY_ID = "urn:oid:2.16.840.1.113883.3.9621";
+
+function constructErrorResponse(
+  payload: PatientDiscoveryRequestIncoming,
+  codingSystem: string,
+  code: string,
+  error: string
+): PatientDiscoveryResponseOutgoing {
+  return {
+    id: payload.id,
+    timestamp: payload.timestamp,
+    responseTimestamp: new Date().toISOString(),
+    patientMatch: null,
+    xcpdHomeCommunityId: METRIPORT_HOME_COMMUNITY_ID,
+    operationOutcome: {
+      resourceType: "OperationOutcome",
+      id: payload.id,
+      issue: [
+        {
+          severity: "error",
+          code: "processing",
+          details: {
+            coding: [{ system: codingSystem, code: code }],
+            text: error,
+          },
+        },
+      ],
+    },
+  };
+}
+
+function constructNoMatchResponse(
+  payload: PatientDiscoveryRequestIncoming
+): PatientDiscoveryResponseOutgoing {
+  return {
+    id: payload.id,
+    timestamp: payload.timestamp,
+    responseTimestamp: new Date().toISOString(),
+    patientMatch: false,
+    xcpdHomeCommunityId: METRIPORT_HOME_COMMUNITY_ID,
+    operationOutcome: {
+      resourceType: "OperationOutcome",
+      id: payload.id,
+      issue: [
+        {
+          severity: "error",
+          code: "processing",
+          details: {
+            text: "no match found",
+          },
+        },
+      ],
+    },
+  };
+}
+
+function constructMatchResponse(
+  payload: PatientDiscoveryRequestIncoming,
+  patient: Pick<Patient, "id" | "data">
+): PatientDiscoveryResponseOutgoing {
+  return {
+    id: payload.id,
+    timestamp: payload.timestamp,
+    responseTimestamp: new Date().toISOString(),
+    patientMatch: true,
+    externalGatewayPatient: {
+      id: patient.id,
+      system: "000", // TBD	- what is this
+    },
+    patientResource: convertPatientToFHIR(patient),
+    xcpdHomeCommunityId: METRIPORT_HOME_COMMUNITY_ID,
+  };
+}
+
+export async function processIncomingRequest(
+  payload: PatientDiscoveryRequestIncoming,
+  // workaround to allow injecting the behavior since this is an isolated function, not a class
+  mpi: MPI
+): Promise<PatientDiscoveryResponseOutgoing> {
+  try {
+    const patient = validateFHIRAndExtractPatient(payload.patientResource);
+    const matchingPatient = await mpi.findMatchingPatient(patient);
+    if (!matchingPatient) {
+      return constructNoMatchResponse(payload);
+    }
+    return constructMatchResponse(payload, patientMPIToPartialPatient(matchingPatient));
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  } catch (error: any) {
+    switch (error.constructor) {
```

Comment:
I think we should stick to `instanceof`, even if it doesn't allow us to use `switch`. See how using `.constructor` would break if we want to check based on the hierarchy (extends): [link](https://jsfiddle.net/fxubq69a/3/).

Just gotta check types from the most specific to the most abstract.

Reply:
great catch. will do this from now on, since it also means I dont need to disable linting